src/proto/streams/send.rs

```diff
@@ -81,28 +80,61 @@ where
         Ok(())
     }
 
+    /// Send an RST_STREAM frame
+    ///
+    /// # Arguments
+    /// + `clear_queue`: if true, all pending outbound frames will be cleared,
+    ///    if false, the RST_STREAM frame will be appended to the end of the
+    ///    send queue.
     pub fn send_reset(
         &mut self,
         reason: Reason,
         stream: &mut store::Ptr<B, P>,
         task: &mut Option<Task>,
+        clear_queue: bool,
     ) {
+        let is_reset = stream.state.is_reset();
+        let is_closed = stream.state.is_closed();
+        let is_empty = stream.pending_send.is_empty();
+        trace!(
+            "send_reset(..., stream={:?}, ..., clear_queue={:?}); \
+             is_reset={:?}; is_closed={:?}; pending_send.is_empty={:?}; \
+             state={:?} \
+            ",
+            stream.id,
+            clear_queue,
+            is_reset,
+            is_closed,
+            is_empty,
+            stream.state
+        );
         if stream.state.is_reset() {
             // Don't double reset
+            trace!(
+                " -> not sending RST_STREAM ({:?} is already reset)",
+                stream.id
+            );
             return;
         }
 
         // If closed AND the send queue is flushed, then the stream cannot be
-        // reset either
-        if stream.state.is_closed() && stream.pending_send.is_empty() {
+        // reset explicitly, either. Implicit resets can still be queued.
+        if is_closed && (is_empty || !clear_queue) {
```

Comment:
It took me a bit to figure out this logic. Would it be simpler to move the `clear_queue` bit before doing this check? Then you can keep the original check (is_closed && is_empty).

Reply:
@seanmonstar suggested this conditional, but I agree it's not super clear. I can rewrite this.

Reply:
Not critical either way. I think the logic is correct.